tls/autotls-client.md

```diff
@@ -0,0 +1,396 @@
+- [References](#references)
+
+
+
+## Overview
+Most modern web browsers only establish TLS connections with peers that present certificates issued by a recognized Certificate Authority (CA).
+Self-signed certificates are generally not accepted.
+To obtain a CA-issued certificate, a requester must complete an ACME (Automatic Certificate Management Environment) challenge.
+This typically involves provisioning a DNS TXT record on a domain the requester controls.
+
+However, most libp2p peers do not own or control domain names, making it impractical for them to complete DNS-based ACME challenges and, by extension, obtain trusted TLS certificates.
+This limitation hinders direct communication between libp2p peers and standard web browsers.
+
+[AutoTLS](https://blog.libp2p.io/autotls/) addresses this problem by introducing an AutoTLS broker — a server that controls a domain and facilitates ACME challenges on behalf of libp2p peers.
+A peer can request the AutoTLS broker to fulfill an ACME DNS challenge on its behalf.
+Once the broker sets the appropriate DNS record, the requesting peer proceeds to notify the ACME server.
+The ACME server validates the challenge against the broker's domain, and if successful, issues a valid certificate.
+
+This mechanism allows libp2p peers to obtain CA-issued certificates without needing to possess or manage their own domain names.
+
+## General Flow
+The following is the general flow of a successful certificate request and subsequent issuance using AutoTLS.
+Here, "node" refers to the machine running a libp2p peer and requesting the challenge,
+while "broker" and "AutoTLS broker", which are used interchangeably, refer to the server that will fulfil the ACME challenge on behalf of the node.
+
+1. Node requests a challenge from the ACME server.
+2. Node sends the challenge to the broker.
+3. Broker tests node and sets DNS record (fulfilling challenge).
+4. Node queries DNS until it sees that the broker has fulfilled the challenge.
+5. Node signals to ACME server that challenge is fulfilled.
+6. ACME server checks challenge in broker.
+7. Node sends CSR to finalize certificate request.
+8. Node polls ACME server until certificate is ready for download.
+9. Node downloads certificate.
+
+## Parameters
+
+| Parameter                | Description                                                      | Reasonable Default |
+|--------------------------|------------------------------------------------------------------|--------------|
+| `max_dns_retries` | The maximum number of DNS queries that the node SHOULD make before giving up | ???  |
+| `max_dns_timeout` | The maximum number of seconds a node SHOULD wait for DNS records to be set | ???  |
+| `max_acme_poll_retries` | The maximum number of GET requests that the node SHOULD issue to ACME server before giving up | ???  |
+| `max_acme_timeout` | The maximum number of seconds a node SHOULD wait for an ACME resource status to change | ???  |
```

Comment:
Let's Encrypt recommendation for challenge completion is ~10minutes, p2p-forge client aborts sooner, around 3 minutes I think